src/hooks/ContractReadValue.ts

```diff
@@ -0,0 +1,97 @@
+import { Contract } from '@ethersproject/contracts'
+import { useCallback, useState } from 'react'
+import useDeepCompareEffect from 'use-deep-compare-effect'
+import { callContractRead } from 'utils/callContractRead'
+import { contractToRead } from 'utils/contractToRead'
+
+/**
+ * Reads a `contract` value from `contracts`, using the `functionName`.
+ *
+ * Upon load, the contract is initially read. From there, `refetchValue` must be
+ * called.
+ *
+ * @returns refetchValue - Call to refetch and hydrate the `value`.
+ * @returns value - Value returned from the contract.
+ * @returns loading - Whether the contract is currently being read.
+ */
+export function useContractReadValue<C extends string, V>({
+  contract,
+  contracts,
+  functionName,
+  args,
+  version,
+  formatter,
+  valueDidChange,
+}: {
+  version: string
```

Comment:
i wonder if we really need this `version` arg, since we're only using it for logging. there may also be version-agnostic reader instances (i.e. price of ETH, or generic ERC20 calls)

Reply:
Yeah, agree. I will work up in the stack to converge the two into a single shared function